MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java

```diff
@@ -758,14 +786,133 @@ public static Entity getInfantryEntity(UnitGeneratorParameters params, int skill
         return createEntityWithCrew(params.getFaction(), skill, campaign, ms);
     }
     
+    /**
+     * Fill the given transport entity with a bunch of units that it can carry.
+     * Currently only works for infantry transports.
+     * @param transport
+     * @param params
+     * @param skill
+     * @param campaign
+     */
+    private static List<Entity> fillTransport(AtBScenario scenario, Entity transport, UnitGeneratorParameters params, int skill, Campaign campaign) {
+        List<Entity> transportedUnits = new ArrayList<>();
+        
+        // if we've already filled the transport, no need to do it again.
+        if(scenario.getTransportLinkages().containsKey(transport.getExternalIdAsString())) {
+            return transportedUnits;
+        }
+        
+        for(Transporter bay : transport.getTransports()) {
+            if(bay instanceof TroopSpace) {
+                double bayCapacity = ((TroopSpace) bay).getUnused();
+                
+                UnitGeneratorParameters newParams = params.clone();
+                newParams.clearMovementModes();
+                newParams.setUnitType(UnitType.INFANTRY);
+                newParams.setWeightClass(AtBDynamicScenarioFactory.UNIT_WEIGHT_UNSPECIFIED);
+                
+                // to save ourselves having to re-generate a bunch of infantry for smaller bays (3 tons and lower)
+                // we will limit ourselves to generating low-weight foot platoons
+                if(bayCapacity <= IUnitGenerator.FOOT_PLATOON_INFANTRY_WEIGHT) {
+                    newParams.getMovementModes().add(EntityMovementMode.INF_LEG);
+                    newParams.setFilter(inf -> inf.getTons() <= IUnitGenerator.FOOT_PLATOON_INFANTRY_WEIGHT);
+                } else {
+                    newParams.getMovementModes().addAll(IUnitGenerator.ALL_INFANTRY_MODES);
+                    newParams.setFilter(inf -> inf.getTons() <= bayCapacity);
+                }
+                
+                MechSummary ms = campaign.getUnitGenerator().generate(newParams);
+                
+                if (ms == null) {
+                    continue;
+                }
+                
+                Entity infantry = createEntityWithCrew(params.getFaction(), skill, campaign, ms);
+                
+                // if we're dealing with a *really* small bay, drop the # squads down until we can fit it in
+                while(infantry.getWeight() > bayCapacity) {
+                    ((Infantry) infantry).setSquadN(((Infantry) infantry).getSquadN() - 1);
```

Comment:
Could it happen that ANY amount of infantry would be too much for a bay? Does setSquadN allow 0?

Reply:
Theoretically possible, and yes. Added safety check.